lib/core/src/Ouroboros/Network/Client/Wallet.hs

```diff
@@ -475,62 +479,99 @@ type LocalStateQueryResult state = Either AcquireFailure state
 --        │       ┌───────────┴───┐             ┌────────┴───────┐
 --        └───────┤   Acquired    │────────────>│   Querying     │
 --                └───────────────┘             └────────────────┘
---
+-- FIXME: I think the diagram is wrong. It is possible to query multiple times
+-- while we're acquired.
 localStateQuery
-    :: forall m block . (MonadThrow m, MonadSTM m)
-    => TQueue m (LocalStateQueryCmd block m)
+    :: forall m block . (MonadThrow m, MonadSTM m, MonadTimer m, Eq (Point block))
+    => m (Point block)
+    -> TQueue m (LocalStateQueryCmd block m)
         -- ^ We use a 'TQueue' as a communication channel to drive queries from
         -- outside of the network client to the client itself.
         -- Requests are pushed to the queue which are then transformed into
         -- messages to keep the state-machine moving.
     -> LocalStateQueryClient block (Point block) (Query block) m Void
-localStateQuery queue =
+localStateQuery getTip queue =
     LocalStateQueryClient clientStIdle
   where
     clientStIdle
         :: m (LSQ.ClientStIdle block (Point block) (Query block) m Void)
-    clientStIdle = awaitNextCmd <&> \case
-        CmdQueryLocalState pt query respond ->
-            LSQ.SendMsgAcquire pt (clientStAcquiring query respond)
+    clientStIdle = do
+        cmd <- awaitNextCmd
+        pt <- getTip
+        pure $ LSQ.SendMsgAcquire pt (clientStAcquiring cmd pt 0)
 
     clientStAcquiring
-        :: forall state. Query block state
-        -> (LocalStateQueryResult state -> m ())
+        :: LocalStateQueryCmd block m
+        -> Point block
+        -> Int
         -> LSQ.ClientStAcquiring block (Point block) (Query block) m Void
-    clientStAcquiring query respond = LSQ.ClientStAcquiring
-        { recvMsgAcquired = clientStAcquired query respond
+    clientStAcquiring qry oldPt retries = LSQ.ClientStAcquiring
+        { recvMsgAcquired = clientStAcquired qry
         , recvMsgFailure = \failure -> do
-                respond (Left failure)
+            if retries < 1000
+            then do
+                let go = do
+                        newPt <- getTip
+                        if oldPt == newPt
+                        -- FIXME: Not the most elegant approach of waiting for a
+                        -- new tip.
+                        then threadDelay 0.01 >> go
+                        else do
+                            pure $ LSQ.SendMsgAcquire newPt (clientStAcquiring qry newPt (retries + 1))
+                go
+            else do
+                --respond (Left failure)
+                -- TODO: Fail
```

Comment:
I don't think clients should have to care about acquire failures. If we can't get a tip acquired with infinite retries, then something is truly wrong.

(Should in that case have infinite retries though)